src/sidebar/components/thread-list-omega.js

```diff
@@ -0,0 +1,246 @@
+    scrolledThreadId.current = undefined;
+    scrolledThreadOffset.current = -1;
+  };
+
+  // Retrieve an estimate for the y-offset (pixels) of a given thread in
+  // the list of all threads by summing the (possibly-estimated) heights
+  // of the threads above it.
+  const getYOffsetFor = threadId => {
+    const threadIds = topLevelThreads.map(t => t.id);
+    const threadIndex = threadIds.indexOf(threadId);
+    if (threadIndex === -1) {
+      return 0;
+    }
+    return threadIds
+      .slice(0, threadIndex)
+      .reduce(
+        (offset, tid) =>
+          offset + (threadHeights.current[tid] || THREAD_DEFAULT_HEIGHT),
+        0
+      );
+  };
+
+  // If we have recently attempted to scroll to the position of a specific
+  // thread in the thread list, check to see if we have a different calculated
+  // y-offset for that thread than we did at last render: that indicates that
+  // the previous scroll position/y-offset wasn't entirely accurate and the
+  // scroll position should be adjusted.
+  const updateScrollPosition = () => {
+    if (scrolledThreadId.current) {
+      const updatedEstimate = getYOffsetFor(scrolledThreadId.current);
+      if (updatedEstimate !== scrolledThreadOffset.current) {
+        threadEl.current.scrollTop = updatedEstimate;
+        scrolledThreadOffset.current = updatedEstimate;
+      } else {
+        clearScrolledThread();
+      }
+    }
+  };
+
+  // Scroll the `threadEl` such that thread with `id` is at the top.
+  const scrollToThread = id => {
+    setScrolledThread(id);
+    updateScrollPosition();
+  };
+
+  // Some or all of the cached height values for thread cards may be
+  // approximations or the real heights may have changed. Obtain the current
+  // rendered heights for the visible threads and update the cache.
+  const updateThreadHeights = () => {
+    const currentHeights = { ...threadHeights.current };
+    visibleThreads.forEach(visibleThread => {
+      const calculatedHeight = getElementHeightWithMargins(visibleThread.id);
+      if (calculatedHeight) {
+        currentHeights[visibleThread.id] = calculatedHeight;
+      }
+    });
+    threadHeights.current = currentHeights;
+  };
+
+  // When a new annotation is created in the application, scroll to that
+  // annotation's thread (unless it is a highlight or reply)
+  $rootScope.$on(events.BEFORE_ANNOTATION_CREATED, (event, annotation) => {
+    if (!isHighlight(annotation) && !isReply(annotation)) {
+      clearSelection();
+      scrollToThread(annotation.$tag);
+    }
+  });
+
+  useEffect(() => {
+    const scrollEl = threadEl.current;
+    scrollEl.addEventListener('scroll', recalculate);
+    window.addEventListener('resize', recalculate);
+
+    updateThreadHeights();
+    updateScrollPosition();
+    recalculate();
+
+    return () => {
+      scrollEl.removeEventListener('scroll', recalculate);
+      window.removeEventListener('resize', recalculate);
+      recalculate.cancel();
+    };
+  });
```

Comment:
Related to this, I thought we had either a runtime check provided by `preact/debug` or an ESLint rule that checked for `useEffect` without a dependency list, since this is something that is typically only used in rare cases AFAIK.